# RetailSense : Explainable Multi-Horizon Retail Demand Forecasting
## Notebook 03 · Data Quality Assessment & Health Audit

> **Organization:** Celebal Technologies &nbsp;|&nbsp; **Intern:** Ayush Choudhary  
> **Domain:** Data Science | Time Series Forecasting | ML | Explainable AI  
> **Dataset:** Kaggle Store-Item Demand Forecasting (913,000 records · 10 Stores · 50 Items · 5 Years)  

---

### 1. Objective
Perform a rigorous data health audit across all **500 time series** before any feature engineering:
- **Missing value audit** — check all columns for null records
- **Duplicate record validation** — verify uniqueness of (date, store, item) combinations
- **Temporal continuity check** — confirm no calendar gaps across 1,826 days
- **Zero-sales inspection** — identify anomalous zero-sales days
- **Outlier detection** — flag statistical outliers using per-series IQR analysis

---


# RetailSense : Explainable Multi-Horizon Retail Demand Forecasting
## Notebook 03 - Data Quality Assessment & Data Health Audit

> **Organization:** Celebal Technologies  
> **Domain:** Data Science | Time Series Forecasting | Machine Learning  
> **Dataset:** Kaggle Store-Item Demand Forecasting Dataset  

---

### 1. Objective
Perform a rigorous data quality assessment across all 500 time series:
- Missing value audit & duplicate record verification.
- Time-series temporal continuity validation (checking for missing calendar days).
- Zero-sales observation inspection.
- Outlier detection via Store-Item grouping and Interquartile Range (IQR) analysis.



In [1]:
import sys
from pathlib import Path
sys.path.append('..')

import pandas as pd
from src.data.loader import load_raw_data
from src.data.quality import assess_data_quality, detect_outliers_iqr

train_df, test_df = load_raw_data()
quality_report = assess_data_quality(train_df)

print("=== Enterprise Data Quality Audit Report ===")
for metric, val in quality_report.items():
    print(f"{metric:25s}: {val}")

Memory usage optimized from 27.86 MB to 10.45 MB (62.5% reduction).
Memory usage optimized from 1.37 MB to 0.60 MB (56.2% reduction).
=== Enterprise Data Quality Audit Report ===
total_records            : 913000
expected_records         : 913000
missing_values           : {'date': 0, 'store': 0, 'item': 0, 'sales': 0}
duplicate_records        : 0
unique_stores            : 10
unique_items             : 50
total_series             : 500
date_range               : ('2013-01-01', '2017-12-31')
total_days               : 1826
continuity_status        : Pass
zero_sales_count         : 1
zero_sales_percent       : 0.0


### 2. IQR Outlier Detection Analysis

In [2]:
outliers_df = detect_outliers_iqr(train_df)
print(f"Total IQR Outliers Flagged Across Series: {len(outliers_df):,} ({len(outliers_df)/len(train_df)*100:.2f}% of total data)")
display(outliers_df.head(10))

Total IQR Outliers Flagged Across Series: 4,522 (0.50% of total data)


,date,store,item,sales
159,2013-06-09,1,1,38
886,2015-06-06,1,1,43
914,2015-07-04,1,1,40
926,2015-07-16,1,1,41
928,2015-07-18,1,1,38
929,2015-07-19,1,1,38
936,2015-07-26,1,1,38
991,2015-09-19,1,1,40
1230,2016-05-15,1,1,39
1249,2016-06-03,1,1,38


### Data Health Summary
The audit confirms 100% date continuity with zero missing values and 1 zero-sales anomaly. Dataset health is verified for Exploratory Data Analysis.

---
### Data Quality Audit Summary

| Check | Result | Status |
|---|---|---|
| Total Records | 913,000 | ✅ Complete |
| Expected Records (500 × 1826) | 913,000 | ✅ Match |
| Missing Values | 0 across all columns | ✅ Clean |
| Duplicate (date, store, item) rows | 0 | ✅ No duplicates |
| Unique Stores | 10 | ✅ Verified |
| Unique Items | 50 | ✅ Verified |
| Date Range | 2013-01-01 → 2017-12-31 | ✅ Complete |
| Temporal Continuity | **PASS** — No gaps detected | ✅ Pass |
| Zero-Sales Anomalies | 1 record (0.00%) | ✅ Negligible |
| IQR Outliers Flagged | 4,522 records (0.50%) | ⚠️ Minor — seasonal peaks |

**Conclusion:** Dataset is **production-ready** with excellent data integrity. The 4,522 IQR outliers correspond to genuine seasonal demand surges (e.g., summer peaks, holidays) — not data errors. No imputation or correction required.

